File: 10038 - Jolly Jumpers/jollyjumper.cpp
/* @JUDGE_ID:   1318761   10038     CPP   */

/* @BEGIN_OF_SOURCE_CODE */

#include <iostream>
#include <string>

int main() {
    bool isJumper = true;
    int size = 0;
    int diff = 0;

    int jumpers[3001]{0};
    int testers[3001]{0};

    std::size_t space = 0;
    // Assume exit condition is blank line
    while (true) {
        std::string input = "";
        std::getline(std::cin, input);

        if (input == "") {
            break;
        }

        space = 0;

        isJumper = true;

        // get size
        size = std::stoi(input, &space);
        input = input.substr(space);

        // Check size is in-bounds (not needed, but force of habit)
        if (size > 3000) {
            size = 3000;
        }

        // This loop breaks down string into array of numbers
        for (auto i = 0; space <= input.size(); i++) {
            jumpers[i] = std::stoi(input, &space);
            input = input.substr(space);
        }

        // This loop finds the difference of each successive pair
        // of items in the arr- Loop starts at index 0, goes to
        // size - 1 to avoid accessing OOB. For each loop, we set
        // the value to one at a given index = diff
        for (auto i = 0; (i < size - 1) && isJumper ; i++) {
            diff = jumpers[i] - jumpers[i+1];
            // abs
            if (diff < 0) {
                diff = diff * -1;
            } // else{do_nothing()}

            // If the value at testers[diff] is not zero, then
            // the sequence has a duplicate diff- and as it has
            // n numbers, and must have one of every number, then
            // it is, by definition, not a jolly number.
            if (testers[diff] == 0) {
                testers[diff] = 1;
            } else {
                isJumper = false;
                break;
            }
        }

        // This loop determines if the sequence was a jolly number- 
        // Skipping zero, we check every item in the array up to the
        // size of the sequence. If any value is zero, then it is not
        // a jolly number.
        for (auto i = 1; (i < size) && isJumper; i++) {
            if (testers[i] == 0) {
                isJumper = false;
                break;
            } else {
                /*do_nothing()*/
            }
        }

        // Print result for this loop
        if (isJumper) {
            std::cout << "Jolly\n" << std::flush; 
        } else {
            std::cout << "Not jolly\n" << std::flush;
        }

        // Clear our arrays for the next loop.
        for (auto i = 0 ; i <= 3000 ; i++) {
            jumpers[i] = 0;
            testers[i] = 0;
        }

    }
    return 0;
}

/* @END_OF_SOURCE_CODE */

File: 10038 - Jolly Jumpers/jollyjumper.pas
program jollyjumper(input, output);
uses
    sysUtils, StrUtils; // This is mostly useful for string Ops.
type
    jumperArray = array[0..3001] of longInt;
var
    size : longInt;
    diff : longInt;
    counter : longInt; // Re-usable counter variable.
    isJumper : boolean;
    cont : boolean;
    space : sizeInt; // Holds location of the next space character.
    line : ansistring; // Holds a copy of our current line.
    jumpers : jumperArray;
    testers : jumperArray;

    function delim (str : string) : longInt;
    var
        result, counter : longInt;
    begin
        counter := 1;
        while counter < High(str) do
        begin
	    if str[counter] = ' ' then
            begin
                result := counter;
                counter := High(str);
            end;
	    counter := counter + 1;
        end;
        delim := result;
    end;


begin
    cont := true;
    isJumper := true;
    while cont = true do
    begin
        // Read stdin to a string.
        readln(line);
        // If input is empty, then exit out of main loop.
        if (line = '') then
	        cont := false
        else
        begin
            line := Trim(line);
            // First value is the number of inputs- Extract it, and continue
            //size := ExtractSubstr(line, space, ' ');
            space := delim(line);
            size := StrToInt(Copy(line, 0, space - 1));
	    Delete(line, 1, space);
            line := Trim(line);
        end;
        
        for counter := Low(testers) to High(testers) do
        begin
            testers[counter] := 0;
            jumpers[counter] := 0;
        end;
        // If string is non-empty, then we continue, otherwise, we exit.
        if (Length(line) > 0) and (cont = true) then
        begin
            isJumper := true;
            // This loop breaks the string down into an array of numbers
            counter := Low(jumpers);
            while Length(line) > 0 do
            begin
                // get next space
                space := delim(line);
                // Saves the contents of the string after the delimiter to
                //   second
                jumpers[counter] := StrToInt(LeftStr(line, space - 1));
                // Increase the counter, so next loop we save to a different
                //   index
                counter := counter + 1;
                // Deletes the contents of the string after, and including, the
                //   delimiter
                Delete(line, 1, space);
                // Trim whitespace, as needed
                line := Trim(line);
            end; // loop ends when line is empty.

        // This loop finds the difference of each successive pair of items in
        //   the array. The loop starts at index 0, and goes to size - 1 to
        //   avoid accessing out-of-bounds memory. For each loop, we set the
        //   value to one at a given index = diff.
        counter := Low(jumpers);
        while (counter < size - 1) and isJumper do
        begin
            diff := jumpers[counter] - jumpers[counter + 1];
            // abs
            if diff < 0 then
                diff := diff * -1;
            // else do_nothing;

            // If the value at testers[diff] is not zero, then the sqeuence has
            // a duplicate diff, and as it has n numbers, and must have one of
            // every number, then it is, by definition, not a jolly number.
            if testers[diff] = 0 then
                testers[diff] := 1
            else
            begin
                counter := size - 1; // effectively breaks loop
                isJumper := false;
            end;
	    counter := counter + 1; // increment every loop.
        end;

        // This loop determines if the sequence was a jolly number- Skipping
        //   zero, we check every item in the array up to the size of the
        //   sequence. If any value is zero, then it is not a jolly number.
        counter := 1;
        while (counter < size) and (isJumper) do
        begin
            if testers[counter] = 0 then
            begin
                isJumper := false;
                counter := size;
            end;
	    counter := counter + 1; // increment every loop.
        end;


        // Print result for this loop.
        if isJumper then
            writeln('Jolly')
        else
            writeln('Not jolly');

        // reset

        for counter:= Low(testers) to High(testers) do
        begin
            testers[counter] := 0;
            jumpers[counter] := 0;
        end;
    end;
end;
end.
